ducktape/tests/session_context.py

```diff
@@ -83,4 +134,75 @@ def generate_results_dir(session_id):
     :type session_id: str
     :rtype: str
     """
-    return session_id + "-test-results"
\ No newline at end of file
+    return session_id + "-test-results"
+
+
+def mkdir_p(path):
+    """mkdir -p functionality.
+    :type path: str
+    """
+    try:
+        os.makedirs(path)
+    except OSError as exc: # Python >2.5
+        if exc.errno == errno.EEXIST and os.path.isdir(path):
+            pass
+        else:
+            raise
+
+
+class TestContext(Logger):
+    """Wrapper class for state variables needed to properly run a single 'test unit'."""
+    def __init__(self, session_context, module, cls, function, config, log_config=None):
+        """
+        :type session_context: ducktape.tests.session_context.SessionContext
+        """
+        self.module = module
+        self.cls = cls
+        self.function = function
+        self.config = config
+        self.session_context = session_context
+
+        self.results_dir = os.path.join(self.session_context.results_dir, self.cls.__name__)
+        mkdir_p(self.results_dir)
+
+        self._logger_configured = False
+        self.configure_logger(log_config)
+
+    @property
+    def test_id(self):
+        name_components = [
+            self.session_context.session_id,
+            self.module]
+
+        if self.cls is not None:
+            name_components.append(self.cls.__name__)
+
+        name_components.append(self.function.__name__)
+        return ".".join(name_components)
+
+    @property
+    def logger_name(self):
+        return self.test_id
+
+    def configure_logger(self, log_config=None):
+        if self._logger_configured:
+            raise RuntimeError("test logger should only be configured once.")
+
+        self.logger.setLevel(logging.DEBUG)
+
+        mkdir_p(self.results_dir)
+        fh = logging.FileHandler(os.path.join(self.results_dir, "test_log"))
+        fh.setLevel(logging.DEBUG)
+        # create console handler with a higher log level
+        ch = logging.StreamHandler(sys.stdout)
+        ch.setLevel(logging.INFO)
+        # create formatter and add it to the handlers
+        formatter = logging.Formatter('[%(levelname)-6s - %(asctime)s - %(module)s - %(funcName)s - lineno:%(lineno)s]: %(message)s')
```

Comment:
Was this formatting constant the same as the one before? Should it only be defined in one location that would be reasonable for everyone to use (e.g. a SessionContext class field, since SessionContext is for globals anyway? or SessionContext instance field, which can be used here since it has self.session_context)?